lib/Dialect/FIRRTL/Transforms/InnerSymbolDCE.cpp

```diff
@@ -0,0 +1,120 @@
+
+#define DEBUG_TYPE "firrtl-inner-symbol-dce"
+
+using namespace mlir;
+using namespace circt;
+using namespace firrtl;
+using namespace hw;
+
+struct InnerSymbolDCEPass : public InnerSymbolDCEBase<InnerSymbolDCEPass> {
+  void runOnOperation() override;
+
+private:
+  void findInnerRefs(Attribute attr);
+  void insertInnerRef(InnerRefAttr innerRef);
+  void removeInnerSyms(FModuleOp op);
+
+  DenseSet<std::pair<StringAttr, StringAttr>> innerRefs;
+};
+
+/// Find all InnerRefAttrs inside a given Attribute.
+void InnerSymbolDCEPass::findInnerRefs(Attribute attr) {
+  // Check if this Attribute is an InnerRefAttr.
+  if (auto innerRef = dyn_cast<InnerRefAttr>(attr))
+    insertInnerRef(innerRef);
+
+  // Check if any sub-Attributes are InnerRefAttrs.
+  if (auto subElementAttr = dyn_cast<SubElementAttrInterface>(attr))
+    subElementAttr.walkSubAttrs([&](Attribute subAttr) {
+      if (auto innerRef = dyn_cast<InnerRefAttr>(subAttr))
+        insertInnerRef(innerRef);
+    });
+}
+
+/// Add an InnerRefAttr to the set of all InnerRefAttrs.
+void InnerSymbolDCEPass::insertInnerRef(InnerRefAttr innerRef) {
+  StringAttr moduleName = innerRef.getModule();
+  StringAttr symName = innerRef.getName();
+
+  auto [iter, inserted] = innerRefs.insert({moduleName, symName});
+  if (!inserted)
+    return;
+
+  ++numSymbolsFound;
+
+  LLVM_DEBUG(llvm::dbgs() << DEBUG_TYPE << ": found " << moduleName
+                          << "::" << symName << '\n';);
+}
+
+/// Remove all InnerSymAttrs within the FModuleOp that are dead code.
+void InnerSymbolDCEPass::removeInnerSyms(FModuleOp moduleOp) {
+  // Walk all ops in the FModuleOp.
+  moduleOp.walk([&](Operation *op) {
+    // Check if the op has an InnerSymAttr.
+    auto innerSym = op->getAttrOfType<InnerSymAttr>("inner_sym");
+    if (!innerSym)
+      return;
+
+    assert(moduleOp == op->getParentOfType<FModuleOp>() &&
+           "ops with inner_sym must be inside an FModuleOp");
+
+    // Check if the InnerSymAttr was found as part of any InnerRefAttr.
+    auto moduleName = moduleOp.moduleNameAttr();
+    auto symName = innerSym.getSymName();
+    if (innerRefs.contains({moduleName, symName}))
+      return;
+
+    // If not, it's dead code.
+    op->removeAttr("inner_sym");
+
+    ++numSymbolsRemoved;
+
+    LLVM_DEBUG(llvm::dbgs() << DEBUG_TYPE << ": removed " << moduleName
+                            << "::" << symName << '\n';);
+  });
+}
+
+void InnerSymbolDCEPass::runOnOperation() {
+  // Run on the top-level ModuleOp to include any VerbatimOps that aren't
+  // wrapped in a CircuitOp.
+  ModuleOp topModule = getOperation();
+
+  // Traverse the entire IR once.
+  SmallVector<FModuleOp> modules;
+  topModule.walk([&](Operation *op) {
+    // Find all InnerRefAttrs.
+    for (NamedAttribute namedAttr : op->getAttrs())
+      findInnerRefs(namedAttr.getValue());
+
+    // Collect all FModuleOps.
+    if (auto moduleOp = dyn_cast<FModuleOp>(op))
+      modules.push_back(moduleOp);
+  });
```

Comment:
I was trying to think of a way that this could only visit the top-level stuff and not do a deep walk. However, this is clearly covering the situation where there could be uses _anywhere_ as opposed to just in hierpath ops.

Reply:
Yeah... I was thinking about that too. I think we need to visit all the top-level stuff, and also go as deep as FModuleOp to collect them so we can iterate in parallel after the initial walk. Maybe once we can do a pre-order walk and stop once we get to FModuleOp?

Reply:
Yeah I think this works, I will push a small update. It now visits the builtin module, then everything in it (verbatim/circuit), then the things inside the circuit (hierpaths and modules), and nothing else.

Reply:
Ooh actually, this doesn't work, and was caught by a unit test. For data taps, we may have verbatim expr ops in the bodies of modules that may contain inner refs, so it does need to descend into the body of modules at least for now. In tests on real designs, the whole IR walk wasn't prohibitively expensive (this pass was no slower than any other pass that visits the whole IR), so I think I'll just keep this as-is.